---
title: "OncoSimulR_Simulation's-Recipes"
author: "David Moreno Gambín, Deyanira Borroto Alburquerque, Guillermo Cerillo Sánchez, Rubén Rincón Álvarez"
date: "9th January 2025"
output: html_document
bibliography: OncoSimulR-rmd.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# OncoSimulR

Cancer can be defined as a group of diseases caused by the abnormal growth of cells in an organ or system, which can simultaneously affect other regions of the body. This process dramatically impacts vital equilibrium.

From different fields, this phenomenon is being studied to be deciphered. For instance, mathematics and informatics are leveraged to understand the complex landscapes of cancer diseases and tumour evolution.

In this context, a software tool called **OncoSimulR**, an R package, was developed. Based on stochastic models, it incorporates different kinds of mutations: passenger mutations (neutral in terms of fitness) and driver mutations (those that enhance tumour fitness).

OncoSimulR is primarily used to simulate the behavior of asexual reproduction in cancer progression. It is a highly flexible tool, allowing users to account for and adjust various parameters. A diversity of evolutionary scenarios and features can be replicated, such as epistasis (gene-to-gene interaction), clonal competition, and different fitness models. For example:

-   McFarland model: To study the balance between beneficial and deleterious mutations.
-   Bozic model: Incorporating growth dynamics and mutation rates.

Additionally, OncoSimulR supports scalable simulations with functions like OncoSimulIndiv and OncoSimulPop, which can simulate either an individual tumour or a population of tumours.

It is also possible to use the mclapply function, enabling parallel simulations, making this tool highly efficient for large-scale studies.

It is important to note that to obtain different simulations with the software, the fitness or evolutionary fitness parameter is used.

## Fitness

In cancer biology, fitness represents the ability of a cell (or genotype) to survive, proliferate, and pass on its genetic material. Higher fitness means a genotype is more likely to dominate the population[@Nowell1976].

### Fitness in OncoSimulR:

Fitness in the context of cancer evolution refers to the capacity of a genotype to outcompete others in the tumour microenvironment. It encapsulates **proliferation rate**, where faster-dividing cells have a higher fitness; **survival**, where cells that resist apoptosis or evade immune detection are fitter; and **adaptation**, which refers to the ability of cells to adapt to hypoxia or therapy, these will present with selective advantages[@osr].

Fitness effects determine how advantageous or disadvantageous mutations are. They are the core of OncoSimulR's simulations. It's important to make them simple but flexible, to be able to simulate many different situations. You define fitness values based on genotypes or specific mutations. Fitness landscapes can be additive (fitness is summed across mutations) or epistatic (interactions between mutations influence fitness),[@osr].

<!-- “Monotypic population” means that for virtually all the time (except for very brief periods of fast clonal sweeps) there is a single genotype, defined with respect to the mutations being modelled, [@apg2024]. …a clone is fully replaced by another clone, where clone is defined with respect to the mutations of relevance; in other words, for the mutations of relevance, the population is monotypic (also from “A picture guide…”) In cancer progression there are two key terms with the power of shaping the tumour dynamics, [@apg2024]. -->


## Strong/weak mutation and strong/weak selection

Strong and weak mutation refers to a regime or landscape where the frequency of mutation rate can be strong, like in late stage cancer tumours where genetic imbalance is dangerous or weak mutation where the mutation rate is low.

In OncoSimulR you can adjust functions, in specific, mu to be equal to a concrete mutation rate, 1e-6 would be a weak one and 1e-4 would be a strong one.

Strong and weak selection refers to a regime or landscape where fitness differences between genotypes are large and beneficial mutations are fixed immediately whereas in a weak selection regime the differences between different fitness are negligible or not large.

In OncoSimulR you can adjust functions to simulate these conditions, in particular, with fitness effect with values 0.1 to strong selection and 0.01 to weak selection.

## Intra-tumour Heterogeneity (ITH)

*"Cancer is the result of a gradual accumulation of somatic genetic mutations. While most of the acquired mutations are putatively neutral and have no significant effect on a cell’s phenotype, some confer a selective advantage to the host cell; they are known as driver mutations. Consequently, individual tumours are heterogeneous and typically consist of multiple populations of cells (subclones), each harboring a distinct set of driver mutations and possessing a distinct phenotype, a phenomenon known as intra-tumour heterogeneity (ITH). Detecting ITH helps identify the key events initiating the development of the disease or leading to metastasis, and allows for the determination of a tumour’s subclonal composition."* [@k2019].

### Characteristics
#### Diverse origins and types
**Intra-tumour heterogeneity** (ITH) refers to the coexistence of genetically, epigenetically, and phenotypically diverse cell populations within a single tumour. Ecompassing all forms of variation within a tumour (including genetic mutations, copy number alterations, epigenetic modifications and differential gene expression). This diversity arises due both neutral processes, such as random mutations, and selective pressures, where microenvironmental selection pressures during tumour development such as competition for nutrients and immune evasion play a big role [@ith2017; @ith2018].

ITH can be observed also as **spatial and temporal heterogeneity**. **Spatial ITH** refers to a situation where different regions of the tumour harbor distinct clones due to localized selection pressures. On the other hand, **temporal ITH** refers to a scenario where the tumour’s clonal composition changes over time as new mutations accumulate and selective pressures shift [@ith2012; @ithNR2018].

#### Clinical implications of ITH
This heterogeneity has consequences for diagnosis, treatment and disease progression.

**High ITH** allows tumours to adapt to changing environments, such as therapy or immune response, therefore it is associated with treatment resistance, immune evasion and poor prognosis. Furthermore, it causes diverse subpopulations to cooperate or compete, influencing tumour growth and metastasis [@Morris2016].


### Modeling ITH in OncoSimulR
The following scenarios are using the McFL model, which is based on the McFarland model of tumor growth. The most important characteristics of this model are the following:
- Fitness-dependent Growth: Each genotype grows at a rate proportional to its fitness relative to the population's average fitness.
- Competition: There is explicit competition among clones for shared resources or space.
- Population Size Dynamics: The total population size is not fixed but fluctuates based on the relative fitness of clones.
- Carrying Capacity: There is an implicit assumption of limited resources, which creates a competitive environment among clones.

Since our objective was to simulate ITH with coexistence of different genotypes within the same environment, we decided to use the McFl model because it adds competition between genotypes and models growth dynamics as a function of fitness. 

#### Scenario 1: Varying fitness

Starting from a random fitness landscape with few genotypes (3 genotypes) and low variations in fitness (therefore, fitness is almost neutral):

```{r}
library(OncoSimulR)
set.seed(1)

ab1 <- rfitness(3, scale = c(1.2, 0, 1))
feab1 <- allFitnessEffects(genotFitness = ab1)

evalAllGenotypes(feab1)
```

Simulating tumor progression with `oncoSimulIndiv` on the created `allFitnessEffects` (feab1). The most relevant parameters here are:
  
-   mutation rate (`mu`): set high
-   initial size (`initSize`): set on the low side, and
-   final time: set high in order to see if results sustain as time passes.

```{r}
out1 <- oncoSimulIndiv(feab1,
                       model = "McFL",
                       onlyCancer = FALSE,
                       finalTime = 2000,
                       mu = 1e-3, 
                       initSize = 1e4)

out1
```

Plotting the results:

```{r}
plot(out1, show = "genotypes", addtot = TRUE, type = "line", thinData = TRUE,
     plotDiversity = TRUE, legend = FALSE)
```

```{r out1, echo=FALSE}
plot(out1, show = "genotypes", addtot = TRUE, type = "line", thinData = TRUE,
     plotDiversity = TRUE, legend = FALSE)
```

**Interpretation:**

Simulations with HIGH mutation rate 1st there is a peak then the population settles and remains stable in 3 genotypes one kind of dominant and another two at lower ammounts - Is this low heterogeneity?

You observe a stable population with one dominant genotype and two minor genotypes. This can be considered low ITH because the population is not evenly distributed among multiple genotypes. The dominance of one genotype reduces overall diversity. You can confirm this by calculating diversity indices (e.g., Shannon's entropy or Simpson's diversity index). A low value of Shannon's entropy would indicate low heterogeneity.
Plotting the results:

```{r}
plot(out1, show = "genotypes", addtot = TRUE, type = "line", thinData = TRUE,
     plotDiversity = TRUE, legend = FALSE)
```

**Interpretation:**

The Diversity plot (the small plot on the top) represents Shannon's entropy, which indicates the diversity indices within the population by time. We will use this plot to interpret the results of the simulation with regards to intra-tumour heterogeneity. First we observe a peak in heterogeneity, then the population settles and we can see a stable population with one dominant genotype and two minor ones. <br> This can be considered low intra-tumour heterogeneity (ITH) because the population is not evenly distributed among multiple genotypes. The dominance of one genotype reduces overall diversity. In fact, in the Diversity plot we can observe the Shannon's entropy settles in a low value (less than 0.2) indicating low diversity indices and therefore low heterogeneity. 



## Tunnelling:
A very visual way to show what happens to our simulations when we stray away from the SSWM regime (and thus, our population is no longer monotypic) is the appearance of “prohibited” genotypes, understanding “prohibited” as those genotypes that are not possible according to our DAG.
If we were, say, to model a scenario as: WT → A → B → C, then “B”, “C”, “A, C” and “B, C” would be prohibited genotypes. The only possible genotypes according to our DAG would be “WT”, “A”, “A, B” and “A, B, C”.
If we run 1000 simulations under a SSWM regime, the chance of finding these genotypes in our population is very low; however, when we change the parameters and move away from SSWM, these prohibited genotypes appear almost always.
Could this be due to the stochastic tunnelling effect?
Stochastic tunnelling (STUN) is a phenomenon in which a step with low fitness is skipped. For example, in a scenario as “WT → A → B” where the fitness of A is lower than WT and the fitness of B is higher, a population of cancer cells could be able to tunnel from WT to B without ever reaching A. This description is paraphrased from Iwasa et al (2004).
Under deterministic models and SSWM regimes, either tunnelling is not possible, or the order restrictions have been violated (as it appears to be the case). It can, however, happen under SSSM or WSSM regimes. Also, STUN is not a problem to consider under stochastic methods.
> I’m not sure how to check what methods OncoSimul is using…
> Maybe it uses all and makes a consensus, idk
> DAG: WT -> A -> B -> C

my_DAG_3 <- allFitnessEffects(
  data.frame(parent = c("Root", "A"),
         	child = c("A", "B"),
         	s = c(-0.1, 0.5),
         	sh = 0.0,
         	typeDep = "MN"),
  drvNames = c("A", "B"))

In this DAG the fitness of “A” is 0.9 and the fitness of “B” is 1.0, because sh is set to 0.0, so the simulation, under not SSWM, could go from “WT” to “A, B”. Assuming that we used keepPhylog = TRUE, the plot from plotClonePhylog() would show that the Line of Descent (LOD) is actually WT → B → A, B. In a monotypic population (SSWM), however), “A, B” won’t probably ever happen, because A is deleterious and tunneling is improbable.

The LOD is actually stored as a df in simulation[["other"]][["LOD_DF"]]. As it can be seen upon inspection, in non-monotypic populations there is a row in which the parent is “B” and the child is “A, B”. This could be categorized as tunneling, for the correct LOD according to our DAG should be “A” → “A, B”, which is not present. In fact, in the monotypic population neither of the formerly mentioned appear, as “B” → “A, B” would violate the order restrictions and “A” is deleterious. However, “WT” → “B” does appear in these monotypic populations, but it never reaches fixation.

However, this is possible because “B” is not deleterious compared to “A”. A more strict definition of STUN would require “WT” → “A, B” without “A” or “B” ever happening. That is, that the Path of the mAximum (POM) and the LOD should not be the same.
First we need to make B also deleterious, turning sh into a negative value. Upon inspecting plot(EvalAllGenotypes(DAG)), The POM is “WT” → “A, B” because “A” and “B” are now both deleterious. If we run the simulations 1000 times for both monotypic and non-monotypic and then inspect all their LODs, “WT” → “A, B” does not occur ever.

Thus, we could conclude that stochastic tunneling, as  

## References
