---
title: "A recipe book for Simulations with OncoSimulR"
author: "David Moreno Gambín, Deyanira Borroto Alburquerque, Guillermo Cerillo Sánchez, Rubén Rincón Álvarez"
date: "9th January 2025"
output: html_document
bibliography: OncoSimulR-rmd.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error = TRUE, silent = TRUE)
```

# OncoSimulR

Cancer can be defined as a group of diseases caused by the abnormal growth of cells in an organ or system, which can simultaneously affect other regions of the body. This process dramatically impacts vital equilibrium.

From different fields, this phenomenon is being studied to be deciphered. For instance, mathematics and informatics are leveraged to understand the complex landscapes of cancer diseases and tumour evolution.

In this context, a software tool called **OncoSimulR**, an R package, was developed. Based on stochastic models, it incorporates different kinds of mutations: passenger mutations (neutral in terms of fitness) and driver mutations (those that enhance tumour fitness).

OncoSimulR is primarily used to simulate the behavior of asexual reproduction in cancer progression. It is a highly flexible tool, allowing users to account for and adjust various parameters. A diversity of evolutionary scenarios and features can be replicated, such as epistasis (gene-to-gene interaction), clonal competition, and different fitness models. For example:

-   McFarland model: To study the balance between beneficial and deleterious mutations.
-   Bozic model: Incorporating growth dynamics and mutation rates.

Additionally, OncoSimulR supports scalable simulations with functions like OncoSimulIndiv and OncoSimulPop, which can simulate either an individual tumour or a population of tumours.

It is also possible to use the mclapply function, enabling parallel simulations, making this tool highly efficient for large-scale studies.

It is important to note that to obtain different simulations with the software, the fitness or evolutionary fitness parameter is used.

## Fitness

In cancer biology, fitness represents the ability of a cell (or genotype) to survive, proliferate, and pass on its genetic material. Higher fitness means a genotype is more likely to dominate the population[@Derbal2023].

### Fitness in OncoSimulR:

Fitness in the context of cancer evolution refers to the capacity of a genotype to outcompete others in the tumour microenvironment. It encapsulates **proliferation rate**, where faster-dividing cells have a higher fitness; **survival**, where cells that resist apoptosis or evade immune detection are fitter; and **adaptation**, which refers to the ability of cells to adapt to hypoxia or therapy, these will present with selective advantages[@osr].

Fitness effects determine how advantageous or disadvantageous mutations are. They are the core of OncoSimulR's simulations. It's important to make them simple but flexible, to be able to simulate many different situations. You define fitness values based on genotypes or specific mutations. Fitness landscapes can be additive (fitness is summed across mutations) or epistatic (interactions between mutations influence fitness),[@osr].




## Strong/weak mutation and strong/weak selection

By taking into account the mutation rate and its selection effect, a model can be classified into different regimes considering whether there is a **strong** or a **weak mutation** and whether there is a **strong** or **weak selection**. 

Strong and weak mutation refers to a regime or landscape where the frequency of mutation rate can be strong, like in late stage cancer tumours where genetic imbalance is dangerous, or weak mutation, where the mutation rate is low. 

Strong and weak selection refers to a regime or landscape where fitness differences between genotypes are large and beneficial mutations are fixed immediately. This means there is a genotype that predominates among all the different genotypes in a population.  On the other hand, in a weak selection regime the differences between different fitness are negligible or not large. Thus, there will be many different mutant lineages in the population at the same time. 

Whenever there is a strong selection, and a tumour population consists of a single genotype, we can refer to the population as a monotypic population, where all the cells share the same mutations.

In the **Strong Selection Weak Mutation (SSWM) regime**, the positive selection of a mutation results in a succession of selective sweeps. This means that whenever a mutation occurs, it rapidly spreads through the population and ends fixing. Some time later another beneficial mutation may occur, resulting in the new genotype replacing the previous one. This regime is typical of small populations in which beneficial mutations are rare. Also in small populations the predominance of a new genotype is favoured by events such as drift. 

In OncoSimulR, by using a small population, SSWM can be simulated as follows: 

```{r OncoSimulR, echo=FALSE}
library(OncoSimulR)
```

```{r}
set.seed(1)
r7b <- rfitness(7, scale = c(1.2, 0, 1))

(sr7b <- oncoSimulIndiv(allFitnessEffects(genotFitness = r7b),
                        model = "McFL",
                        mu = 1e-6,
                        onlyCancer = FALSE,
                        finalTime = 40000,
                        initSize = 1e3,
                        keepEvery = 4,
                        detectionSize = 1e10))
plot(sr7b, show = "genotypes", legend = FALSE)
```
We can observe how the WT genotype is fastly substituted by a new genotype, that predominates though a long time until another beneficial mutation appears and replaces the previous genotype. 

When the population increases its size, its more likely that beneficial mutations appear more often, resulting in many genotypes coexisting at the same time, distancing ourselves from the SSWM regime. 

This is a similar example as the previous one, but with a bigger population. 
```{r}
set.seed(1)
r7b <- rfitness(7, scale = c(1.2, 0, 1))

(sr7b <- oncoSimulIndiv(allFitnessEffects(genotFitness = r7b),
                        model = "McFL",
                        mu = 1e-6,
                        onlyCancer = FALSE,
                        finalTime = 400,
                        initSize = 1e7,
                        keepEvery = 4,
                        detectionSize = 1e10))
plot(sr7b, show = "genotypes", legend = FALSE)
```
As we can observe, there are several different subpopulations with different mutations. 

However, over time there can be strong selection. It can happen that there is a moderately beneficial mutation, not strong enough to sweep the entire population, so that later a stronger mutation takes place and overcomes the previous mutation. Another possible cause is that a lineage suffers multiple mutations, being the first one beneficial, but is thanks to the effect of the following mutations that this lineage is strongly selected. 

We can simulate this example just by increasing the time in our simulation: 
```{r}
set.seed(1)
r7b <- rfitness(7, scale = c(1.2, 0, 1))

(sr7b <- oncoSimulIndiv(allFitnessEffects(genotFitness = r7b),
                        model = "McFL",
                        mu = 1e-6,
                        onlyCancer = FALSE,
                        finalTime = 4000,
                        initSize = 1e7,
                        keepEvery = 4,
                        detectionSize = 1e10))
plot(sr7b, show = "genotypes", legend = FALSE)
```

This now belongs to the Strong Selection Strong Mutation (SSSM) regime. 

## Intra-tumour Heterogeneity (ITH)

*"Cancer is the result of a gradual accumulation of somatic genetic mutations. While most of the acquired mutations are putatively neutral and have no significant effect on a cell’s phenotype, some confer a selective advantage to the host cell; they are known as driver mutations. Consequently, individual tumours are heterogeneous and typically consist of multiple populations of cells (subclones), each harboring a distinct set of driver mutations and possessing a distinct phenotype, a phenomenon known as intra-tumour heterogeneity (ITH). Detecting ITH helps identify the key events initiating the development of the disease or leading to metastasis, and allows for the determination of a tumour’s subclonal composition."* [@k2019].

### Characteristics
#### Diverse origins and types
**Intra-tumour heterogeneity** (ITH) refers to the coexistence of genetically, epigenetically, and phenotypically diverse cell populations within a single tumour. Ecompassing all forms of variation within a tumour (including genetic mutations, copy number alterations, epigenetic modifications and differential gene expression). This diversity arises due both neutral processes, such as random mutations, and selective pressures, where microenvironmental selection pressures during tumour development such as competition for nutrients and immune evasion play a big role [@ith2017; @ith2018].

ITH can be observed also as **spatial and temporal heterogeneity**. **Spatial ITH** refers to a situation where different regions of the tumour harbor distinct clones due to localized selection pressures. On the other hand, **temporal ITH** refers to a scenario where the tumour’s clonal composition changes over time as new mutations accumulate and selective pressures shift [@ith2012; @ithNR2018].

#### Clinical implications of ITH
This heterogeneity has consequences for diagnosis, treatment and disease progression.

**High ITH** allows tumours to adapt to changing environments, such as therapy or immune response, therefore it is associated with treatment resistance, immune evasion and poor prognosis. Furthermore, it causes diverse subpopulations to cooperate or compete, influencing tumour growth and metastasis [@Morris2016].


### Modeling ITH in OncoSimulR
The following scenarios are using the McFL model, which is based on the McFarland model of tumor growth. The most important characteristics of this model are the following:

- Fitness-dependent Growth: Each genotype grows at a rate proportional to its fitness relative to the population's average fitness.
- Competition: There is explicit competition among clones for shared resources or space.
- Population Size Dynamics: The total population size is not fixed but fluctuates based on the relative fitness of clones.
- Carrying Capacity: There is an implicit assumption of limited resources, which creates a competitive environment among clones.

Since our objective was to simulate ITH, and therefore, coexistence of different genotypes within the same environment, we decided to use the McFl model because it adds competition between genotypes and models growth dynamics as a function of fitness. 

------------------------------------------------------------------------

#### Modeling variations in fitness:

------------------------------------------------------------------------

#### Scenario 1: Low fitness variations

Starting from a random fitness landscape with few genotypes (3 genotypes) and low variations in fitness (therefore, fitness is almost neutral):

```{r}
set.seed(1)

ab1 <- rfitness(3, scale = c(1.1, 0, 1))
feab1 <- allFitnessEffects(genotFitness = ab1)

evalAllGenotypes(feab1)
```

Simulating tumor progression with `oncoSimulIndiv` on the created `allFitnessEffects` (feab1). The most relevant parameters here are:

-   mutation rate (`mu`): set high,
-   initial size (`initSize`): set at an intermediate amount, and
-   `finalTime`: set long enough (500) to see how the population stabilizes

Therefore, we are creating a scenario of **weak selection strong mutation (WSSM)**.

```{r}
out1 <- oncoSimulIndiv(feab1,
                       model = "McFL",
                       onlyCancer = FALSE,
                       finalTime = 500,
                       mu = 1e-3, 
                       initSize = 1e5)
```

**Interpretation:**

We can see the final population composition by running `out1`:

```{r out1, echo=FALSE}
out1
```

The population consists of one dominant genotype (C) and another three established genotypes but with lower populations.

By plotting the results we can obtain more information:

```{r}
plot(out1, show = "genotypes", addtot = TRUE, plotDiversity = TRUE,
     thinData = TRUE, type = "line", legend.ncols = 4, ylim = c(1,1e8))
```

The Diversity plot (referring to the small plot on the top) represents **Shannon's entropy**, which indicates the diversity indices within the population by time. We will use this plot to interpret the results of the simulations with regards to intra-tumour heterogeneity. 

First we observe a peak in heterogeneity, then the population settles and we can see a stable population with one dominant genotype and three minor ones. This can be considered **low intra-tumour heterogeneity** (ITH) because the population is not evenly distributed among multiple genotypes and the dominance of one genotype reduces overall diversity. In fact, in the Diversity plot we can observe the Shannon's entropy settles in a low value (around 0.2) indicating low diversity indices.


#### Scenario 2: Low fitness variations

This is the same scenario as before, but with 4 genotypes and very high mutation rates:

```{r}
set.seed(1)

ab7 <- rfitness(4, scale = c(1.1, 0, 1))
feab7 <- allFitnessEffects(genotFitness = ab7)
evalAllGenotypes(feab7)
out7 <- oncoSimulIndiv(feab7,
                       model = "McFL",
                       onlyCancer = FALSE,
                       finalTime = 500,
                       mu = 1e-2, 
                       initSize = 1e5)
```

**Interpretation:**

```{r}
out7
```

This is a scenario of **weak selection strong mutation (WSSM)** that yields a high number of coexisting genotypes and high diversity (\~2.0). 

First we observe the usual peak in populations, however this time most of the populations settle at a similar number. This means that multiple genotypes are in coexistence which indicates **high ITH**.

```{r}
plot(out7, show = "genotypes", addtot = TRUE, plotDiversity = TRUE,
     type = "line", thinData = TRUE, legend.ncols = 4, 
     ylim = c(1,1e10))
```

#### Scenario 3: High fitness variations

With high fitness variations between genotypes we cannot get coexistence and therefore is not possible to reach intra-tumour heterogeneity (ITH).

It happens as it should, high fitness differences between all the genotypes results in a **monotypic population** (where most of the time there is a single genotype) and this is incompatible with ITH. This occurs because high fitness differences causes that the fitter one overrides the others... To cause ITH then we should try to move towards a **non-monotypic population**.

To do this, at least two genotypes must have similar high fitness. Therefore, we need a number of genotypes and some of them with similarly high fitness.

```{r}
set.seed(180)
ab3 <- rfitness(4, scale = c(10, 5, 1))
feab3 <- allFitnessEffects(genotFitness = ab3)
evalAllGenotypes(feab3)
out3 <- oncoSimulIndiv(feab3,
                       model = "McFL", 
                       onlyCancer = FALSE,
                       finalTime = 50,
                       mu = 1e-2, 
                       initSize = 1e2)
```

We can see that the fitness at birth of A and B and A,C are rather high.

```{r}
out3
```

In fact, we can see that A yields the largest population; followed by A,C; and then A,D and A,B.

```{r}
plot(out3, show = "genotypes", addtot = TRUE, plotDiversity = TRUE,
     type = "line", legend.ncols = 4, ylim = c(1,1e12))
```

These results show a low to moderate sustained diversity (\~0.5).

------------------------------------------------------------------------

#### Aiming high ITH

------------------------------------------------------------------------

#### Scenario 1

In this scenario a very weak selection and many different interfering genes allow for mutations to thrive and coexist.

```{r}
set.seed(149)
genes1 <- rfitness(6, scale = c(1, 1.02))

drivergenes <- which(genes1 > 0)
fx1 <- allFitnessEffects(genotFitness = genes1)

evalAllGenotypes(fx1)

sim1 <- oncoSimulIndiv(fx1,
                       model = "McFL",
                       onlyCancer = FALSE,
                       finalTime = 2000,
                       mu = 1e-3, 
                       initSize = 1e4)
```

This represents high sustained ITH. We can observe in the diversity plot constant rates over 2.0, and reaching 3.0 at some points.

```{r}
plot(sim1, show = "genotypes", addtot = TRUE, 
     plotDiversity = TRUE, thinData = TRUE, legend = FALSE)
```

#### Scenario 2: High ITH with many clones
In this scenario, genomic instability causes multiple mutations to accumulate, starting from one mutator phenotype with strong selection and leading to many subclones each with unique mutations that allow them all to coexist.

First creating 9 genes with no interaction and another 9 with ranging low interaction:

```{r}
set.seed(1)

genes <- c(rep(0, 9), runif(9, -0.01, 0.1)) 
names(genes) <- c("a1", "a2", "a3", "b1", "b2", "c1", "c2", "c3", "c4",
                  paste0("h", 1:9))

hgenes <- which(genes > 0) # separate genes with positive fitness effects
```

`drvNames` marks the genes with positive fitness effects (`hgenes`) as driver genes (for later specification during plotting):
                      
```{r}
fx <- allFitnessEffects(noIntGenes = genes,
                        drvNames = names(genes)[hgenes])
```

Then defining `allMutatorEffects` will influence mutation rates:

```{r}
mx <- allMutatorEffects(epistasis = c("A" = 3,
                                      "B" = 5,
                                      "C" = 15,
                                      "A:B" = 50),
                        geneToModule = c("A" = "a1, a2, a3",
                                         "B" =  "b1, b2",
                                         "C" = "c1, c2, c3, c4")) 
set.seed(2)
sim_ith <- oncoSimulIndiv(fx, muEF = mx,
                          mu = 1e-6, # baseline mutation rate
                          initSize = 1e6,
                          model = "McFL",
                          onlyCancer = FALSE,
                          finalTime = 1100,
                          keepPhylog = TRUE, # keep the genealogy of the clones
                          detectionSize = NA,
                          detectionProb = NA,
                          detectionDrivers = NA,
                          errorHitWallTime = FALSE,
                          errorHitMaxTries = FALSE,
                          seed = NULL)
```

```{r}
plot(sim_ith, plotDrivers = TRUE, addtot = TRUE, plotDiversity = TRUE,
     thinData = TRUE, legend.ncols = 5, ylim = c(1,1e9))
```

This model creates high and dynamic ITH due to the combination of mutator effects and random fitness effects, this simulation generates high genetic diversity. The presence of mutators (mx) accelerates the generation of new genotypes, maintaining a dynamic population with frequent emergence and competition of clones. In this way the population keeps fluctuating among genotypes.



## Tunnelling (& Valley Crossing)
A very visual way to show what happens to our simulations when we stray away from the SSWM regime (and thus, our population is no longer monotypic) is the frequent appearance of genotypes incompatible with our DAG.

If we were to model a scenario as: WT → A → B → C, then “B”, “C”, “A, C” and “B, C” would be incompatible genotypes. The only possible genotypes according to our DAG would be “WT”, “A”, “A, B” and “A, B, C”. If we run 1000 simulations under a SSWM regime, the chance of finding these genotypes in our population is very low; however, when we change the parameters and move away from SSWM, these incompatible genotypes appear almost always.

You may (or not, who knows) be wondering “why does this happen?” Well, this occurrence might be due, in part, to valley crossing.
A valley is the depression between two elevated terrains; so, in our fitness landscape, it refers to a deleterious genotype situated between two beneficial genotypes. Sometimes, reaching the most beneficial genotype requires going through a deleterious one; this is known as valley crossing. Now, two asexual populations may have different approaches to cross the valley depending on various parameters such as population size, mutation rate and fitness of compatible and incompatible genotypes. So, as could be expected, valley crossing will differ from a monotypic population to a non-monotypic one.

In monotypic populations, under the SSWM regime, we typically have small populations and a reasonable difference of fitness between genotypes, which makes the population much more susceptible to drift. So, the most common case of valley crossing in these scenarios is sequential fixation, which means that all genotypes in the line of descent (LOD) are fixated (sometimes by genetic drift) one after another until the population reaches the fitness summit. For not monotypic populations, as the population sizes are usually much larger, combined with different mutation rates and fitness effects, stochastic tunneling (STUN) is more usual in these particular cases [@Weissman2009].

STUN is the phenomenon in which intermediate genotypes drift at relatively low frequencies, each such intermediate eventually disappearing, until an individual accumulates a combination of mutations that provides a selective advantage. Essentially, it means that a population goes from state 0 to state 2 without ever reaching (fixating, actually) state 1 [@Iwasa2004].

# STUN Parameters in the Literatue
According to [@Iwasa2004], the easiest way is to have a DAG as 0 → 1 → 2 and fitness effects as fe0 = 1, fe1 =< 1, fe2 > 1 (~10); also, the mutation rate for 1 has to be lower that 2 (mu1 = 1e-5, mu2 = 1e-2). However, if we create a DAG like “WT → A → B”, our stage 2 is actually going to be “A, B”, not “B”. In OncoSimul, we can change the fitness effect of “A, B”, but we were not able to change its mutation rate because “A, B” is not a gene in the DAG (it is a genotype, not a gene). So, we won’t keep a difference in mutation rates for “A” and “B”. The key factor according to this paper is that the population size must remain constant, this can be achieved in OncoSimul by setting the model variable to “Cont”.

For this very same DAG (with 2 genes), [@Weissman2009] provide more useful indications, such as how the population size and the selective disadvantage (parameter sh in OncoSimul) for intermediate single-mutants (“A” and “B”) will affect the type of valley crossing that the cell population experiences. What we should take home from this paper is to keep the population above 1/muA and the sh variable close to 0. Again, this paper specifies that mu2 is the mutation rate from the single-mutant to the double-mutant. We are not sure if that is possible in OncoSimul, but haven’t found out how. So, it is quite possible that these models are somewhat different to those of the papers.

# Modeling STUN with OncoSimulR
First, we need to call some packages, as well as setting the number of cores for our parallelized functions depending on which OS we're using.

```{r pakages-cores, message=FALSE, warning=FALSE}
library(OncoSimulR)
library(graph)
library(igraph)
library(parallel)
igraph_options(vertex.color = "SkyBlue2")

if(.Platform$OS.type == "windows") {
  mc.cores <- 1
} else {
  mc.cores <- 2
}
```

Then, we need to create the DAG. For this, we will use the allFitnessEffects function with a data.frame that contains the parental relationships of our genotypes. The variable "s" is the fitness for the genotypes that do not violate the order retrictions, 15.667 is a number tailored to make "A, B"'s fitness roughly equal to 10 (as suggested in the literature. The "sh" variable is the fitness penalty for the genotypes that do not follow the DAG, again inspired by the literature. The type of dependancy is set to "MN" (or "AND"), but does not really matter in this case.

```{r tunneling-DAG}
# DAG: WT -> A -> B
my_DAG <- allFitnessEffects(
  data.frame(parent = c("Root", "A"),
         	child = c("A", "B"),
         	s = c(-0.4, 15.667),
         	sh = -0.004,
         	typeDep = "MN"),
  drvNames = c("A", "B"))
```

Now, we can show different plots for our DAG and our Fintess Landsacpe, as well as a table with the fitness values for all posible combination of genes.

```{r DAG-plots}
plot(my_DAG) #Direct Acyclic Graph
(evag <- evalAllGenotypes(my_DAG, order = FALSE, addwt = TRUE)) #The Fitness of all posible genotypes
plot(evag, use_ggrepel = TRUE) #Fitness Landscape
```

We can now run 1 or many simulations using the mclapply() function in this example, we will run only 100 so that it goes fast, as we won't be going along all of them.
It is fundamental that the model is set to "Const", so that the population size remains constant. The time our simulations will be running can be low, as tunneling will occur pretty fast. The size of our population must be enough to get out of the SSWM regime and the mutation rate should be relatively fast. The variable keepPhylog allow us to conserve the progression of our population and keepEvery is set to TRUE for maximum accuracy.

```{r tunneling-sim}
# Not Monotypic Population
set.seed(6) #For reproducibility
not_mono_pop <- mclapply(1:100, function(x) oncoSimulIndiv(my_DAG, model = "Const",
                                                          	mu = 1e-3,
                                                          	onlyCancer = FALSE,
                                                          	finalTime = 100,
                                                          	keepPhylog = TRUE,
                                                          	keepEvery = TRUE,
                                                            initSize = 1e5), #high pop.
                mc.cores = mc.cores)
```

We can see how our population evolves with the following tables and graphs.
The table shows the number of cells of each genotype. The columns are (from left to right): Time, "WT", "A", "A, B" and "B".
The plot shows the emergence and disappearance of different genotypes as lines, so all can be seen clearly.

```{r tunneling-plots}
head(not_mono_pop[[1]]$pops.by.time) #Population size of each gebotype by time.
plot(not_mono_pop[[1]], show = "genotypes", type = "line") #Grwoth lines for our genotypes.
```

As it can be seen, intermediate single-mutants appear but do not fixate, and the double-mutant fixates quite fast when “A” or “B” appear (with mu = 1e-3 it would be almost intant). Thus, we have STUN.
Stochastic Tunneling is just one way of crossing fitness valleys and there exist some variations of STUN depending on the parameters of our simulations.
One example is Sequential Fixation, which happens when all genotypes of the LOD fixate. This normally happens with neutral valleys but, under very precise circumstances, it can also happen with deleterious valleys just like STUN. The key parameter here is siez, which must be very small. For it to work properly, the mutation rates of "A" and "B" were changes with a named vector. If both were too fast, STUN would happen; if both were too slow, nothing would happen.
We won't enter in more detail as we would run out of time. For more details on STUN, please check [@Iwasa2004] and [@Weissman2009].

# Sequential Fixation
```{r seq-fix}
# DAG: WT -> A -> B
my_DAG <- allFitnessEffects(
  data.frame(parent = c("Root", "A"),
             child = c("A", "B"),
             s = c(-0.02, 2),
             sh = -0.02,
             typeDep = "MN"),
  drvNames = c("A", "B"))

set.seed(1)
mu_rates <- c("A" = 1e-5, "B" = 1e-1)
Seq_Fix <- mclapply(1:100,function(x) oncoSimulIndiv(my_DAG,
                                    model = "Const",
                                    mu = mu_rates,
                                    onlyCancer = FALSE,
                                    finalTime = 1000,
                                    keepPhylog = TRUE,
                                    keepEvery = TRUE,
                                    initSize = 100), #low pop.
                       mc.cores = mc.cores)

(Seq_Fix[[1]]$GenotypesLabels)
(Seq_Fix[[1]]$pops.by.time)
plot(Seq_Fix[[1]], show = "genotypes", type = "line")
plot(Seq_Fix[[1]], show = "genotypes")
```

# Checking our simulations
Now, of course, we should check if our results are sufficient. This is, wi will check the percentage of simulations in which we obtain the desired genotype.
First, we check which simulations have the double-mutant, then, we check if it has reached fixation and, for SeqFix, if the single-mutant has reached fixation.
This may sound a bit sketchy, but it is a quick way to see more or less how many of our simulations reached the desired condition.

# STUN:
STUN_genotypes <- c()
ID <- 0
IDs <- c()
for (sim in STUN) {
  ID <- ID + 1
  STUN_genotype <- c(STUN_genotype, sim[["GenotypesLabels"]])
  if (sim[["GenotypesLabels"]][length(sim[["GenotypesLabels"]])-1]=="A, B") {
    IDs <- c(IDs, ID)
  }
}
rm(sim)
table(STUN_genotypes)

count_AB_fix <- 0
for (ID in IDs) {
  rower <- 0
  gen_labels <- STUN[[ID]]$GenotypesLabels
  span <- STUN[[ID]]$pops.by.time[,length(gen_labels)]
  for (idx in span) {
    if (idx >= 1e4) {rower <- rower + 1}}
  if (rower >= 1) {count_AB_fix <- count_AB_fix + 1}
}
rm(ID)
count_AB_fix

# Seq-Fix:
SeqFix_genotypes <- c()
ID <- 0
IDs <- c()
for (sim in Seq_Fix) {
  ID <- ID + 1
  SeqFix_genotype <- c(SeqFix_genotype, sim[["GenotypesLabels"]])
  if (sim[["GenotypesLabels"]][length(sim[["GenotypesLabels"]])-1]=="A, B") {
    IDs <- c(IDs, ID)
  }
}
rm(sim)
table(SeqFix_genotypes)

count_B_fix <- 0
for (ID in IDs) {
  rower <- 0
  gen_labels <- Seq_Fix[[ID]]$GenotypesLabels
  span <- Seq_Fix[[ID]]$pops.by.time[,length(gen_labels)+1]
  for (idx in span) {
    if (idx >= 90) {rower <- rower + 1}}
  if (rower >= 1) {count_B_fix <- count_B_fix + 1}
}
rm(ID)
count_B_fix
```

The results will vary depending on how the parameters are set ad which seed we're using. As it was mentioned before, valley crossing has many variants and explaining all would take a long time. Hopefuly, this broad painting of the concepts will be sufficient to understand the very basics.

  
## References

## Session Info
```{r sessionInfo}
sessionInfo()
```
